src/frequenz/sdk/actor/power_distributing/power_distributing.py

```diff
@@ -62,6 +64,78 @@ def _get_all(source: dict[int, frozenset[int]], keys: abc.Set[int]) -> set[int]:
     return set().union(*[source[key] for key in keys])
 
 
+def _get_battery_inverter_mappings(
+    batteries: Iterable[int],
+    *,  # force keyword arguments
+    inv_bats: bool = True,
+    bat_bats: bool = True,
+    inv_invs: bool = True,
+) -> dict[str, dict[int, frozenset[int]]]:
+    """Create maps between battery and adjacent inverters.
+
+    Args:
+        batteries: batteries to create maps for
+        inv_bats: whether to create inverter to battery map
+        bat_bats: whether to create battery to battery map
+        inv_invs: whether to create inverter to inverter map
+
+    Returns:
+        a dict of the requested maps, using the following keys:
+            * "bat_invs": battery to inverters map
+            * "inv_bats": inverter to batteries map
+            * "bat_bats": battery to batteries map
+            * "inv_invs": inverter to inverters map
+    """
+    bat_invs_map: dict[int, set[int]] = {}
+    inv_bats_map: dict[int, set[int]] | None = {} if inv_bats else None
+    bat_bats_map: dict[int, set[int]] | None = {} if bat_bats else None
+    inv_invs_map: dict[int, set[int]] | None = {} if inv_invs else None
+    component_graph = connection_manager.get().component_graph
+
+    for battery_id in batteries:
+        inverters: set[int] = set(
+            component.component_id
+            for component in component_graph.predecessors(battery_id)
+            if component.category == ComponentCategory.INVERTER
+        )
+
+        if len(inverters) == 0:
+            _logger.error("No inverters for battery %d", battery_id)
+            continue
+
+        _logger.debug("Battery %d has inverter %s", battery_id, list(inverters))
+
+        bat_invs_map[battery_id] = inverters
+        if bat_bats_map is not None:
+            bat_bats_map.setdefault(battery_id, set()).update(
+                set(
+                    component.component_id
+                    for inverter in inverters
+                    for component in component_graph.successors(inverter)
+                )
+            )
+
+        for inverter in inverters:
+            if inv_bats_map is not None:
+                inv_bats_map.setdefault(inverter, set()).add(battery_id)
+            if inv_invs_map is not None:
+                inv_invs_map.setdefault(inverter, set()).update(bat_invs_map)
+
+    mapping: dict[str, dict[int, frozenset[int]]] = {}
+
+    # Convert sets to frozensets to make them hashable.
+    if bat_invs_map:
+        mapping["bat_invs"] = {k: frozenset(v) for k, v in bat_invs_map.items()}
+    if inv_bats_map:
+        mapping["inv_bats"] = {k: frozenset(v) for k, v in inv_bats_map.items()}
+    if bat_bats_map:
+        mapping["bat_bats"] = {k: frozenset(v) for k, v in bat_bats_map.items()}
+    if inv_invs_map:
+        mapping["inv_invs"] = {k: frozenset(v) for k, v in inv_invs_map.items()}
```

Comment:
what about of defining a helper to convert `set` to `frozenset` as this could be useful for other use cases?

Reply:
sorry, I meant the set of items-values in the dict, anyway I thought this could be useful to avoid code-repetition in other use-cases but it's fine to keep it as it is